core/providers/rust/rust.go

```diff
@@ -0,0 +1,693 @@
+	miseStep := ctx.GetMiseStepBuilder()
+	p.InstallMisePackages(ctx, miseStep)
+	if p.shouldUseMusl(ctx) {
+		miseStep.AddSupportingAptPackage("musl-tools")
+	}
+
+	install := ctx.NewCommandStep("install")
+	install.AddInputs([]plan.Input{
+		plan.NewStepInput(miseStep.Name()),
+	})
+	p.Install(ctx, install)
+
+	build := ctx.NewCommandStep("build")
+	build.AddInputs([]plan.Input{
+		plan.NewStepInput(miseStep.Name()),
+		plan.NewStepInput(install.Name(), plan.InputOptions{
+			Exclude: []string{"/app/"},
+		}),
+	})
+	p.Build(ctx, build)
+
+	maps.Copy(ctx.Deploy.Variables, p.GetRustEnvVars(ctx))
+	ctx.Deploy.Inputs = []plan.Input{
+		ctx.DefaultRuntimeInput(),
+		plan.NewStepInput(miseStep.Name(), plan.InputOptions{
+			Include: miseStep.GetOutputPaths(),
+		}),
+		plan.NewStepInput(build.Name(), plan.InputOptions{
+			Include: []string{"/app/bin"},
+		}),
+	}
+	ctx.Deploy.StartCmd = p.GetStartCommand(ctx)
+
+	return nil
+}
+
+func (p *RustProvider) StartCommandHelp() string {
+	return "To start your Rust application, Railpack will look for:\n\n" +
+		"1. A main.ts, main.js, main.mjs, or main.mts file in your project root\n\n" +
+		"2. If no main file is found, it will use the first .ts, .js, .mjs, or .mts file found in your project\n\n" +
+		"The selected file will be run with `rust run --allow-all`"
+}
+
+func (p *RustProvider) GetStartCommand(ctx *generate.GenerateContext) string {
+	target := p.getTarget(ctx)
+
+	if target != "" {
+		workspace := p.resolveCargoWorkspace(ctx)
+
+		if workspace != "" {
+			return fmt.Sprintf("./bin/%s", workspace)
+		}
+
+		return p.getStartBin(ctx)
+	}
+
+	workspace := p.resolveCargoWorkspace(ctx)
+
+	if workspace != "" {
+		return fmt.Sprintf("./bin/%s", workspace)
+	}
+
+	return p.getStartBin(ctx)
+}
+
+func (p *RustProvider) getStartBin(ctx *generate.GenerateContext) string {
+	bins, err := p.getBins(ctx)
+	if err != nil {
+		return ""
+	}
+
+	if len(bins) == 0 {
+		return ""
+	}
+
+	var bin string
+
+	if len(bins) == 1 {
+		bin = bins[0]
+	} else if envBinName, _ := ctx.Env.GetConfigVariable("RUST_BIN"); envBinName != "" {
+		for _, b := range bins {
+			if b == envBinName {
+				bin = b
+				break
+			}
+		}
+	} else {
+		cargoToml, err := parseCargoTOML(ctx)
+		if err == nil && cargoToml.Package.DefaultRun != "" {
+			bin = cargoToml.Package.DefaultRun
+		}
+	}
+
+	if bin == "" {
+		return bin
+	}
+
+	binSuffix := p.getBinSuffix(ctx)
+	return fmt.Sprintf("./bin/%s%s", bin, binSuffix)
+}
+
+func (p *RustProvider) getTarget(ctx *generate.GenerateContext) string {
+	// Target may be defined in .config/cargo.toml
+	if p.shouldMakeWasm32Wasi(ctx) {
+		return "wasm32-wasi"
+	}
+
+	if p.shouldUseMusl(ctx) {
+		return getRustTarget(ctx)
+	}
+
+	return ""
+}
+
+func (p *RustProvider) Install(ctx *generate.GenerateContext, install *generate.CommandStepBuilder) {
+	ctx.Caches.AddCache("cargo_registry", CARGO_REGISTRY_CACHE)
+	ctx.Caches.AddCache("cargo_git", CARGO_GIT_CACHE)
```

Comment:
The caches need to be added to the specific step. Something like

```go
install.AddCache(ctx.Caches.AddCache("cargo_registry", CARGO_REGISTRY_CACHE))
```

_Realizing the ergonomics here are a bit strange. Lmk if you have any suggestions_